User status: JS tooltips for Buddy list and PM list, changes to full user profile's last seen field.

vinitS101 opened the pull request:
>A cleaner PR that resolves #11607 

### Added JS Tooltips for User circles and user names in the Buddy List (right side bar) and the Private Messages List (left side bar).

1. Hovering on a User Circle reveals Active/Idle/Offline/Unavailable for green/orange/white/white-with-line user circle.
2. Hovering on the Names of Individual users in either list reveals:
    - Name
    - Status Message (if set)
    - Last online (Eg: "Online now" or "Last online: 10 minutes ago")
3. Hovering on Group Names reveals the name of all users in the group.

*Update: Example GIFs in a comment below.*

showell commented on static/js/buddy_data.js:
```diff
@@ -48,6 +41,22 @@ exports.get_user_circle_class = function (user_id) {
     }
 };
 
+exports.get_user_circle_status = function (user_id) {
```
I would call this function something more like `status_description`.  I think having `user_circle` in the name is confusing, since the names here are pretty independent of whether our icon is a circle, square, or some other thing.

vinitS101 replied:
Changed the name to `status_description`

showell commented on static/js/components.js:
```diff
@@ -125,6 +125,28 @@ exports.toggle = function (opts) {
+    elem.tooltip('show');
+};
+
+exports.register_title_tooltip = function (elem, user_id) {
```
I don't think we want code that is this buddy-list-specific to be in components.  The components piece should just be the stuff that is generic, like `register_tooltip` above.  For code that is custom to the buddy list, we can put that in `buddy_list.js`.

vinitS101 replied:
I'll move it. It is also used by the pm list by the way, but yeah one or the other.

showell commented on static/templates/user_presence_row.handlebars:
```diff
@@ -1,11 +1,10 @@
 <li data-user-id="{{user_id}}" class="user_sidebar_entry {{#if num_unread}} user-with-count {{/if}} narrow-filter {{#if faded}} user-fade {{/if}}">
     <div class="selectable_sidebar_block">
-        <span class="{{user_circle_class}} user_circle"></span>
+        <span class="{{user_circle_class}} user_circle user_circle_buddy" data-hover="{{user_circle_status}}"></span>
```
Why do we need `data-hover` here?  If we're going to build the tooltip HTML when we hover this, then all we need is `data-user-id`, which is already available in the parent.

vinitS101 replied:
This is for the user_circle which do not have complex data to display. The `data-hover` are supposed to work in conjunction with the more generic `register_tooltip` function where we can supply the pre-computed hover text directly. 
I left this as is because a lot of the other html title hovers we have right now can be changed this way.

showell commented on static/js/buddy_data.js:
```diff
@@ -174,33 +183,29 @@ exports.user_last_seen_time_status = function (user_id) {
     var person = people.get_person_from_user_id(user_id);
     var my_user_status = exports.my_user_status(user_id);
-    var title = exports.user_title(user_id);
+    var user_circle_status = exports.get_user_circle_status(user_id);
```
I don't understand why need to add this here.

showell commented on static/js/pm_list.js:
```diff
@@ -108,10 +109,11 @@ exports._build_private_messages_list = function (active_conversation, max_privat
         var user_circle_class = buddy_data.get_user_circle_class(user_ids_string);
-
-        var fraction_present;
+        var user_circle_status = buddy_data.get_user_circle_status(user_ids_string);
```
Why is this code changing?

showell commented:
@timabbott Assuming the first two commits pass tests, they are ready to merge (Rishi gave the lgtm over chat).  And then we can start working through some of the more complex ones.

timabbott commented:
I merged those first two commits, thanks @vinitS101!

vinitS101 commented:
@showell I added the required tests and they seem to work fine. Can you please review this? 
Thanks!

vinitS101 commented:
@timabbott Please review the last 3 commits.

vinitS101 commented:
I just made some styling changes to the tooltips. This PR is pretty much done.
@timabbott @showell Please review this PR when you have the time. Thanks!

vinitS101 commented:
Made the additional changes that @rishig asked for [here](https://github.com/zulip/zulip/issues/11607#issuecomment-474936312).

showell commented on static/js/buddy_data.js:
```diff
@@ -227,11 +215,22 @@ exports.info_for = function (user_id) {
     };
 };
 
+exports.register_title_tooltip = function (elem, user_id) {
```
Why do we call this `register_title_tooltip` instead of `show_buddy_list_title`?  I don't understand where the "register" part comes from.

vinitS101 replied:
I'll make this change ASAP. 
Tim suggested the [name](https://chat.zulip.org/#narrow/stream/91-code-review/topic/.2311882/near/717306) for the function in one of the early reviews. It was for a more generic function though, and I didn't think about changing this when I modified this function to only generate specific tooltips. 
Thanks for pointing it out.

showell commented:
Tim, this is probably ready for a next round of review.

vinitS101 commented:
@timabbott Can you please take a look at this? Thanks!

vinitS101 commented:
Rebased this to latest master.

rishig commented:
Thanks Vinit! Some next steps here design-wise are to 
* Move the popover to the left (for the right sidebar) or the right (for the left sidebar). The positioning can be similar to where we put the stream actions popover.
* For a first pass styling-wise we can just copy the styling (white background, box shadow, etc) of our action popovers
* Potentially it should disappear (or fade out) after 5 seconds 
* right sidebar: There's a dead spot to the left of the down chevron, where the popover doesn't appear on hover.

vinitS101 commented:
> Move the popover to the left (for the right sidebar) or the right (for the left sidebar). The positioning can be similar to where we put the stream actions popover.

Do you want them on the left/right for the user circles as well?

rishig commented:
> Do you want them on the left/right for the user circles as well?

Good question. Let's say 
right sidebar: they should appear on the left
left sidebar: let's just show the same thing we do for the name hover. So left sidebar will only have 1 type of on-hover popover, right sidebar will have two.

vinitS101 commented:
> left sidebar: let's just show the same thing we do for the name hover. So left sidebar will only have 1 type of on-hover popover, right sidebar will have two.

@rishig I am not sure what you mean by this. Do you want me to eliminate the `user_circle` tooltips and just have tooltips appear when hovering on the user name?

> For a first pass styling-wise we can just copy the styling (white background, box shadow, etc) of our action popovers

The actions popover has the default popover styling that bootstrap provides, while the hovers here are `tooltips` which bootstrap styles differently by default. We could have simply used `on-hover popovers` here to make things easier if we wanted this.

> Potentially it should disappear (or fade out) after 5 seconds

I tried this, and it doesn't seem to be working. It could be due to our bootstrap being a very old version (I am not entirely sure). I'll try to fix this in the later stages of this review once I'm done with the other changes.

vinitS101 commented:
Also, if I could suggest something about the tooltip positioning. I think having them at the `bottom` is more inline with their functionality as a replacement for html titles. Having tried a few versions what you suggested, I can't say they look better than the original placement.

rishig commented:
> Do you want me to eliminate the user_circle tooltips and just have tooltips appear when hovering on the user name?

A tooltip should appear when hovering over both the user_circle and the user name. But, it should be the same tooltip. Does that make sense?

> Also, for the user_circles in the right PM list, setting position to left results in them getting hidden in a smaller browser window.

hm, interesting. Roughly how small does the window have to be? 

> I think having them at the bottom is more inline with their functionality as a replacement for html titles.

Sure, we can try this.

vinitS101 commented:
I made these changes.

For the PM List, hovering on the name or user circle show the same user status tooltip.

![pm_list_hover](https://user-images.githubusercontent.com/25329595/62424127-3cc93d00-b6e8-11e9-9350-8b9417d1a0e3.gif)
----------
For Buddy List, hovering over the user names and the user circles yield different tooltips.

![buddy_list_hover](https://user-images.githubusercontent.com/25329595/62424141-57031b00-b6e8-11e9-98a8-6f44f60e69fb.gif)
-----------
The Group PM List has the same behavior as the PM List (since their user circles didn't have any tooltip initially).

![group_pm_list_hover](https://user-images.githubusercontent.com/25329595/62424152-6aae8180-b6e8-11e9-87d2-516e81ae5192.gif)


@rishig

rishig commented:
I think if we're going to show it within the left sidebar, it makes sense to position it similarly to title text? E.g. left-aligned to the mouse, and no triangle.

vinitS101 commented:
Do you want this change for the buddy list as well, or just the left sidebar?

This is what it looks like on the buddy list:
![no_arrow_buddy_list](https://user-images.githubusercontent.com/25329595/62427626-52079100-b713-11e9-8211-56771f40a25d.gif)

--------
Although, there position of the tooltip is stationary if you move the mouse over the hover target.

![problem_buddy_list](https://user-images.githubusercontent.com/25329595/62427634-706d8c80-b713-11e9-89c2-12b56ecaeac2.gif)

rishig commented:
Great, thanks!
* Yeah, lets change it for both left and right sidebar
* Position is stationary if you move the mouse: sounds good
* I guess it's sort of broken for the Group PM case, in that it only shows the info of the first user in the list? Maybe we can just not have this feature for the Group PM section (i.e. no hover behavior). 

One small thing is the text in the hover popover should be left-justified, rather than centered.

vinitS101 commented:
* Removed the tooltip for the Group PMs on the right sidebar.

* Changed the alignment to left.

![Screenshot_3](https://user-images.githubusercontent.com/25329595/62428281-7e280f80-b71d-11e9-9003-5bc3e22e9323.png)

Although, Group names still look like this on the PM List.

![Screenshot_4](https://user-images.githubusercontent.com/25329595/62428300-c0e9e780-b71d-11e9-830e-49e24f936f77.png)

rishig commented:
Great! And using the original popover for the Group PMs makes sense. We should left-align those as well, for consistency.

vinitS101 commented:
@rishig I am not sure what you mean by original popover

vinitS101 commented:
Fixed the alignment issue for group PMs in the PM list.

timabbott commented on frontend_tests/node_tests/buddy_data.js:
```diff
@@ -2,10 +2,15 @@ const _page_params = {};
 zrequire('user_status');
+
+// var noop = function () {};
+// $.fn.tooltip = noop;
```
Clean this up?

vinitS101 replied:
Removed the comments and pushed the changes.

timabbott commented:
I merged the prep commits.

vinitS101 commented on static/templates/group_pms.hbs:
```diff
@@ -7,7 +7,7 @@
         <span class="user_circle_fraction" style="background:none; border-color:hsl(0, 0%, 50%);"></span>
         {{/if}}
-        <a href="{{href}}" data-name="{{name}}" title="{{name}}" class="group-pm-link">{{short_name}}</a>
+        <a href="{{href}}" data-name="{{name}}" class="group-pm-link">{{short_name}}</a>
```
I removed the tooltip for Group PMs on the right sidebar, as suggested, but I've also removed the HTML title that we originally had. Should I leave it here, or is this correct?

vinitS101 replied:
@rishig

vinitS101 commented:
@rishig

vinitS101 commented:
Added a JS tooltip for the Group PMs on the right side bar (similar to the ones in the left sidebar) after discussing it with Rishi.

@timabbott Can you please review this? Thanks!

timabbott commented:
OK, I finally found the time to do a review of this; sorry for the delay.  I made several changes and pushed back to the PR:
* Removed the different hover behavior for the circles; I didn't think it was a good experience in reality.  Ideally we'd just have hovering anything in the region (circle, etc.) other than the chevron trigger the hover widget.
* I fixed a rebase issue with needing to specify `html: true` for these tooltips.
* Extracted some common logic to a new function `do_render_buddy_list_tooltip`.  Could maybe use some further tweaking.
* Made the tooltips a bit wider to handle "Last active: 2 weeks ago" case better.

Some things that still need fixing:
* Node tests are failing since the rebase; not sure what the issue is.
* If you send a PM to a bot, you can get this hover behavior with it to appear in the left sidebar with a weird "Last active: 2 weeks ago" notice.  I think we want to suppress that line for bots and replace it with e.g. "Bot owned by King Hamlet" for bots with an owner, and nothing for bots without one.
* The super-long line selector I added to try to make it cover the whole region needs to be improved; there should be a better way to do hovers over that whole region that's cleaner anyway (and regardless it's a way too long line).  I think there's also a bug in my tweaked version that can make it possible for two tooltips to be visible at once, which seems broken.  Possibly we just want to have the tooltip show code autohide any existing tooltips to address that category?  But we might want to fix this directly first in case it reveals something else we're doing wrong.
* We probably want some sort of debounce to wait 200ms or something before showing the hover content.
* Rishi opened #13261, but we might want to fix that on this branch so we can do the width properly for it.
* The positioning of the tooltips feels a bit off in the left sidebar; it's probably fine to fix-forward on that but just noting it in case it's easy to fix.

@vinitS101 do you have time to tackle those?  I'd love to include this in the 2.1 release.

vinitS101 commented:
I'm just starting work on #13000 . I'll locally check out the changes that you pushed to this PR later tonight. I'll try to wrap both of these in time.

timabbott commented:
Sounds good!

timabbott commented:
@vinitS101 I just noticed you pushed some changes to this 9 days ago; is this ready for another review?  I'd love to get it in for the 2.1 release.

(BTW you should always post a comment when you're ready for another review; otherwise there's no way for maintainers to know whether or not you are)

vinitS101 commented:
I rebased, fixed some minor issues and pushed here to remove conflicts. I didn't leave a comment for review because I haven't been able to test some of the bigger changes I made because of the provisioning issue I described [here](https://chat.zulip.org/#narrow/stream/21-provision-help/topic/provision.20failed/near/793530)

timabbott commented:
OK, good to know.  Hopefully you can resolve that soon.

vinitS101 commented:
> 
> 
> OK, good to know. Hopefully you can resolve that soon.

My dev env has been working fine. Have had to resolve conflicts a bunch of times, but I'm close to finishing up work on this.

vinitS101 commented:
I added the special functionality for bots (with and without owners). Added tests for the same.

![bots_pm_list_hover](https://user-images.githubusercontent.com/25329595/68613304-6f06e980-04e4-11ea-81a3-6e466068ee2c.gif)

timabbott commented:
Not done reviewing, but I've made these changes locally to clean up the code:

```
diff --git a/frontend_tests/node_tests/buddy_data.js b/frontend_tests/node_tests/buddy_data.js
index 02336ecd3f..9b0cbef727 100644
--- a/frontend_tests/node_tests/buddy_data.js
+++ b/frontend_tests/node_tests/buddy_data.js
@@ -128,7 +128,7 @@ run_test('buddy_status', () => {
 
 run_test('title_data', () => {
     // Groups
-    var is_group = 'true';
+    var is_group = true;
     var user_ids_string = "9999,1000";
     var expected_group_data = {
         is_group: is_group,
diff --git a/static/js/buddy_data.js b/static/js/buddy_data.js
index 085e7c9dab..539f0cd3f8 100644
--- a/static/js/buddy_data.js
+++ b/static/js/buddy_data.js
@@ -198,7 +198,7 @@ exports.info_for = function (user_id) {
 };
 
 exports.get_title_data = function (user_ids_string, is_group) {
-    if (is_group === 'true') {
+    if (is_group === true) {
         // For groups, just return a string with recipient names.
         return {
             is_group: is_group,
@@ -227,12 +227,12 @@ exports.get_title_data = function (user_ids_string, is_group) {
         };
     }
 
-    // For buddy list and individual PMS.
+    // For buddy list and individual PMS.  Since is_group=False, it's
+    // a single, human, user.
+    var user_id = user_ids_string;
+    var active_status = presence.get_status(user_id);
     var last_seen = exports.user_last_seen_time_status(user_ids_string);
-    var online_now = false;
-    if (last_seen.indexOf("Active now") !== -1) {
-        online_now = true;
-    }
+    var online_now = active_status === 'active';
 
     return {
         is_group: '',
diff --git a/static/js/click_handlers.js b/static/js/click_handlers.js
index 59360c1246..aeb97999a4 100644
--- a/static/js/click_handlers.js
+++ b/static/js/click_handlers.js
@@ -500,7 +500,7 @@ exports.initialize = function () {
         e.stopPropagation();
         var elem = $(e.currentTarget).closest(".user_sidebar_entry").find(".user-presence-link");
         var user_id = elem.attr('data-user-id');
-        var title_data = buddy_data.get_title_data(user_id, 'false');
+        var title_data = buddy_data.get_title_data(user_id, false);
         do_render_buddy_list_tooltip(elem, title_data);
     });
 
@@ -516,7 +516,8 @@ exports.initialize = function () {
         e.stopPropagation();
         var elem = $(e.currentTarget);
         var user_ids_string = elem.attr('data-user-ids-string');
-        var is_group = elem.attr('data-is-group');
+        // This converts from 'true' in the DOM to true.
+        var is_group = JSON.parse(elem.attr('data-is-group'));
 
         var title_data = buddy_data.get_title_data(user_ids_string, is_group);
         do_render_buddy_list_tooltip(elem, title_data);
```

timabbott commented:
* I also added `i18n.t()` to the active status state strings, so they're properly internationalized.

timabbott commented:
Merged as 329d0126bd7030ad4c35374e6c4b8b913f979504 with those changes, thanks @vinitS101!  There's some i18n follow-up work to be done in #13426.

Huge thanks for making this happen!

vinitS101 commented:
This PR has been my single biggest learning opportunity since I started working on Zulip. It never occurred to me that I could directly use the `===` operator this way. 
@timabbott Thank you for all the help with this PR.
